main/core/src/mill/define/Task.scala

```diff
@@ -106,327 +115,486 @@ trait NamedTask[+T] extends Task[T] {
+   * caching enabled by [[PersistentImpl]]: your [[Worker]] can cache running
+   * sub-processes, JVM Classloaders with JITed code, and all sorts of things
+   * that do not easily serialize to JSON on disk.
+   *
+   * Like [[PersistentImpl]], The user defining a [[Worker]] assumes the
+   * responsibility of ensuring the implementation is idempotent regardless of
+   * what in-memory state the worker may have.
+   */
+  def worker[T](t: Task[T])(implicit ctx: mill.define.Ctx): Worker[T] =
+    macro Internal.workerImpl1[T]
 
-    val taskIsPrivate = isPrivateTargetOption(c)
+  def worker[T](t: Result[T])(implicit ctx: mill.define.Ctx): Worker[T] =
+    macro Internal.workerImpl2[T]
 
-    mill.moduledefs.Cacher.impl0[Target[T]](c)(
-      reify(
-        new TargetImpl[T](
-          t.splice,
-          ctx.splice,
-          rw.splice,
-          taskIsPrivate.splice
-        )
-      )
-    )
-  }
+  /**
+   * Creates an anonymous `Task`. These depend on other tasks and
+   * be-depended-upon by other tasks, but cannot be run directly from the
+   * command line and do not perform any caching. Typically used as helpers to
+   * implement `T{...}` targets.
+   */
+  def task[T](t: Result[T]): Task[T] = macro Applicative.impl[Task, T, mill.api.Ctx]
 
-  def sources(values: Result[os.Path]*)(implicit ctx: mill.define.Ctx): Target[Seq[PathRef]] =
-    macro sourcesImpl1
-
-  def sourcesImpl1(c: Context)(values: c.Expr[Result[os.Path]]*)(ctx: c.Expr[mill.define.Ctx])
-      : c.Expr[Target[Seq[PathRef]]] = {
-    import c.universe._
-    val wrapped =
-      for (value <- values.toList)
-        yield Applicative.impl0[Task, PathRef, mill.api.Ctx](c)(
-          reify(value.splice.map(PathRef(_))).tree
-        ).tree
+  /**
+   * Converts a `Seq[Task[T]]` into a `Task[Seq[T]]`
+   */
+  def sequence[T](source: Seq[Task[T]]): Task[Seq[T]] = new Task.Sequence[T](source)
 
-    val taskIsPrivate = isPrivateTargetOption(c)
+  /**
+   * Converts a `Seq[T]` into a `Task[Seq[V]]` using the given `f: T => Task[V]`
+   */
+  def traverse[T, V](source: Seq[T])(f: T => Task[V]): Task[Seq[V]] = {
+    new Task.Sequence[V](source.map(f))
+  }
 
-    mill.moduledefs.Cacher.impl0[SourcesImpl](c)(
-      reify(
-        new SourcesImpl(
-          Target.sequence(c.Expr[List[Task[PathRef]]](q"_root_.scala.List(..$wrapped)").splice),
-          ctx.splice,
-          taskIsPrivate.splice
-        )
-      )
-    )
+  /**
+   * A variant of [[traverse]] that also provides the [[mill.api.Ctx]] to the
+   * function [[f]]
+   */
+  def traverseCtx[I, R](xs: Seq[Task[I]])(f: (IndexedSeq[I], mill.api.Ctx) => Result[R])
+      : Task[R] = {
+    new Task.TraverseCtx[I, R](xs, f)
   }
 
-  def sources(values: Result[Seq[PathRef]])(implicit ctx: mill.define.Ctx): Target[Seq[PathRef]] =
-    macro sourcesImpl2
+  object Internal {
+    private def isPrivateTargetOption(c: Context): c.Expr[Option[Boolean]] = {
```

Comment:
Shoved all the macro impls into `object Internal` so they don't turn up in Scaladoc. They can't be made private due to macro limitations